File: src/converter/plugins/convertModuleDeclaration.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {NamespaceInfoService, namespaceInfoServiceKey} from "./NamespaceInfoPlugin.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";

export const convertModuleDeclaration = createSimplePlugin((node, context, render) => {
    if (!ts.isModuleDeclaration(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
    const namespaceInfoService = context.lookupService<NamespaceInfoService>(namespaceInfoServiceKey)

    const namespaceStrategy = namespaceInfoService?.resolveNamespaceStrategy(node)

    if (namespaceStrategy === "ignore") {
        return (node.body && render(node.body)) ?? ""
    }

    if (namespaceStrategy === "object") {
        const name = render(node.name)

        const namespace = typeScriptService?.findClosest(node.parent, ts.isModuleDeclaration)

        let externalModifier = "external "

        if (namespace !== undefined && namespaceInfoService?.resolveNamespaceStrategy(namespace) === "object") {
            externalModifier = ""
        }

        const body = (node.body && render(node.body)) ?? ""

        return `
${externalModifier}object ${name} {
${body}
}
        `
    }

    if (namespaceStrategy === "package") {
        // was handled by file structure
        return ""
    }

    throw new Error(`Unknown namespace strategy: ${namespaceStrategy}`)
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertModuleDeclaration.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.configuration.NamespaceStrategy
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isModuleDeclaration

val convertModuleDeclaration = createPlugin plugin@{ node, context, render ->
    if (!isModuleDeclaration(node)) return@plugin null

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
    val namespaceInfoService = context.lookupService<NamespaceInfoService>(namespaceInfoServiceKey)

    val namespaceStrategy = namespaceInfoService?.resolveNamespaceStrategy(node)

    if (namespaceStrategy == NamespaceStrategy.ignore) {
        return@plugin node.body ?.let { render(it) } ?: ""
    }

    if (namespaceStrategy == NamespaceStrategy.`object`) {
        val name = render(node.name)

        val namespace = typeScriptService?.findClosestNamespace(node.parent)

        var externalModifier = "external "

        if (namespace != null && namespaceInfoService.resolveNamespaceStrategy(namespace) == NamespaceStrategy.`object`) {
            externalModifier = ""
        }

        val body = node.body?.let { render(it) } ?: ""

        return@plugin """
${externalModifier}object ${name} {
${body}
}
        """.trim()
    }

    if (namespaceStrategy == NamespaceStrategy.`package`) {
        // was handled by file structure
        return@plugin ""
    }

    error("Unknown namespace strategy: $namespaceStrategy")
}
